File: tests/retention.test.mjs
import assert from "node:assert/strict";
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import test from "node:test";
import { openDatabase } from "../src/db.mjs";
import { createRetention } from "../src/retention.mjs";

const NOW = Date.parse("2026-07-28T12:00:00.000Z");
const isoBefore = (milliseconds) => new Date(NOW - milliseconds).toISOString();
const isoAfter = (milliseconds) => new Date(NOW + milliseconds).toISOString();

async function withRetentionDatabase(fn) {
  const dir = await fs.mkdtemp(path.join(os.tmpdir(), "skytrace-retention-"));
  const dbPath = path.join(dir, "skytrace.db");
  const db = openDatabase(dbPath);
  for (const id of ["rx-1", "rx-2", "rx-3"]) {
    db.prepare("INSERT INTO receivers (id) VALUES (?)").run(id);
  }
  let retention;
  try {
    retention = createRetention({
      dbPath,
      firstRunMs: 3600000,
      now: () => NOW,
      logger: { info() {}, warn() {}, error() {} },
    });
    await fn({ db, retention });
  } finally {
    await retention?.close();
    db.close();
    await fs.rm(dir, { recursive: true, force: true });
  }
}

function insertTrack(db, receiverId, hex, positionAt) {
  db.prepare(`
    INSERT INTO track_points (
      hex, receiver_id, observed_at, position_at, lat, lon
    ) VALUES (?, ?, ?, ?, 37.5, 127.1)
  `).run(hex, receiverId, positionAt, positionAt);
}

function insertCurrent(db, receiverId, hex, observedAt) {
  db.prepare(`
    INSERT INTO receiver_aircraft_current (receiver_id, hex, observed_at)
    VALUES (?, ?, ?)
  `).run(receiverId, hex, observedAt);
}

function insertBatch(db, receiverId, receivedAt) {
  db.prepare(`
    INSERT INTO ingest_batches (
      receiver_id, received_at, aircraft_count, accepted_count
    ) VALUES (?, ?, 0, 0)
  `).run(receiverId, receivedAt);
}

test("retention prunes by timestamps, including receivers without coverage state and future poison", async () => {
  await withRetentionDatabase(async ({ db, retention }) => {
    insertTrack(db, "rx-1", "aaa001", isoBefore(100 * 86400000));
    insertTrack(db, "rx-1", "aaa002", isoBefore(89 * 86400000));
    insertTrack(db, "rx-3", "aaa003", isoBefore(200 * 86400000));
    insertTrack(db, "rx-2", "aaa004", isoAfter(10 * 60000));
    db.prepare(`
      INSERT INTO coverage_receiver_state (
        receiver_id, config_key, origin_lat, origin_lon, last_track_id, updated_at
      ) VALUES ('rx-1', 'test', 37.5, 127, 0, ?)
    `).run(new Date(NOW).toISOString());

    // Insert out of chronological order: deletion must use timestamps, never AUTOINCREMENT ids.
    insertBatch(db, "rx-1", isoBefore(6 * 86400000));
    insertBatch(db, "rx-1", isoBefore(8 * 86400000));
    insertCurrent(db, "rx-1", "bbb001", isoBefore(23 * 3600000));
    insertCurrent(db, "rx-1", "bbb002", isoBefore(25 * 3600000));
    insertCurrent(db, "rx-2", "bbb003", isoAfter(10 * 60000));

    const result = await retention.runNow();
    assert.equal(result.complete, true);
    assert.equal(result.trackOld, 2);
    assert.equal(result.trackFuture, 1);
    assert.equal(result.batches, 1);
    assert.equal(result.currentOld, 1);
    assert.equal(result.currentFuture, 1);
    assert.deepEqual(
      db.prepare("SELECT hex FROM track_points ORDER BY hex").all().map((row) => row.hex),
      ["aaa002"],
    );
    assert.deepEqual(
      db.prepare("SELECT received_at FROM ingest_batches").all().map((row) => row.received_at),
      [isoBefore(6 * 86400000)],
    );
    assert.deepEqual(
      db.prepare("SELECT hex FROM receiver_aircraft_current").all().map((row) => row.hex),
      ["bbb001"],
    );
    assert.equal(result.cursorWarnings.length, 1);
    assert.equal(result.cursorWarnings[0].receiverId, "rx-1");
    assert.deepEqual(result.futureRows.map((row) => row.table).sort(), [
      "receiver_aircraft_current",
      "track_points",
    ]);
    assert.ok(Number.isInteger(result.storage.pageCount));
    assert.ok(Number.isInteger(result.storage.freelistCount));
  });
});

test("a short retention budget resumes in later worker runs and close is idempotent", async () => {
  await withRetentionDatabase(async ({ db, retention }) => {
    const insert = db.prepare(`
      INSERT INTO track_points (
        hex, receiver_id, observed_at, position_at, lat, lon
      ) VALUES (?, 'rx-1', ?, ?, 37.5, 127.1)
    `);
    const old = isoBefore(100 * 86400000);
    db.transaction(() => {
      for (let index = 0; index < 6000; index += 1) {
        insert.run(index.toString(16).padStart(6, "0"), old, old);
      }
    })();

    // Create a second worker with a deliberately tiny tick budget.
    const tiny = createRetention({
      dbPath: db.name,
      firstRunMs: 3600000,
      budgetMs: 1,
      now: () => NOW,
      logger: { info() {}, warn() {}, error() {} },
    });
    try {
      let result;
      for (let run = 0; run < 20; run += 1) {
        result = await tiny.runNow();
        if (result.complete) break;
      }
      assert.equal(result.complete, true);
      assert.equal(db.prepare("SELECT COUNT(*) AS count FROM track_points").get().count, 0);
      await tiny.close();
      await tiny.close();
    } finally {
      await tiny.close();
    }

    // The fixture-owned worker remains independently closable after the peer worker exits.
    assert.equal(retention.state().running, false);
  });
});

File: server/retention/retention_test.go
package retention

import (
	"context"
	"fmt"
	"path/filepath"
	"testing"
	"time"

	"github.com/luftaquila/skytrace/server/database"
)

func TestRunPrunesTimestampsAndFutureRows(t *testing.T) {
	ctx := context.Background()
	db, err := database.Open(ctx, filepath.Join(t.TempDir(), "skytrace.db"))
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()
	for _, receiver := range []string{"rx-1", "rx-2", "rx-3"} {
		if _, err := db.SQL.Exec("INSERT INTO receivers (id) VALUES (?)", receiver); err != nil {
			t.Fatal(err)
		}
	}
	now := time.Date(2026, 7, 28, 12, 0, 0, 0, time.UTC)
	insertTrack := func(receiver, hex string, at time.Time) {
		t.Helper()
		text := iso(at)
		if _, err := db.SQL.Exec(`
			INSERT INTO track_points (
			  hex, receiver_id, observed_at, position_at, lat, lon
			) VALUES (?, ?, ?, ?, 37.5, 127.1)
		`, hex, receiver, text, text); err != nil {
			t.Fatal(err)
		}
	}
	insertTrack("rx-1", "aaa001", now.Add(-100*24*time.Hour))
	insertTrack("rx-1", "aaa002", now.Add(-89*24*time.Hour))
	insertTrack("rx-3", "aaa003", now.Add(-200*24*time.Hour))
	insertTrack("rx-2", "aaa004", now.Add(10*time.Minute))
	if _, err := db.SQL.Exec(`
		INSERT INTO coverage_receiver_state (
		  receiver_id, config_key, origin_lat, origin_lon, last_track_id, updated_at
		) VALUES ('rx-1', 'test', 37.5, 127, 0, ?)
	`, iso(now)); err != nil {
		t.Fatal(err)
	}
	for index, age := range []time.Duration{6 * 24 * time.Hour, 8 * 24 * time.Hour} {
		if _, err := db.SQL.Exec(`
			INSERT INTO ingest_batches (
			  receiver_id, received_at, aircraft_count, accepted_count
			) VALUES ('rx-1', ?, 0, 0)
		`, iso(now.Add(-age))); err != nil {
			t.Fatal(err)
		}
		_ = index
	}
	for index, offset := range []time.Duration{-23 * time.Hour, -25 * time.Hour, 10 * time.Minute} {
		receiver := "rx-1"
		if index == 2 {
			receiver = "rx-2"
		}
		if _, err := db.SQL.Exec(`
			INSERT INTO receiver_aircraft_current (receiver_id, hex, observed_at)
			VALUES (?, ?, ?)
		`, receiver, fmt.Sprintf("bbb%03d", index+1), iso(now.Add(offset))); err != nil {
			t.Fatal(err)
		}
	}
	result, err := Run(ctx, db.SQL, now, Config{
		TrackRetentionDays: 90, BatchRetentionDays: 7, CurrentRetentionHours: 24,
		Budget: time.Second,
	})
	if err != nil {
		t.Fatal(err)
	}
	if !result.Complete || result.TrackOld != 2 || result.TrackFuture != 1 ||
		result.Batches != 1 || result.CurrentOld != 1 || result.CurrentFuture != 1 {
		t.Fatalf("result = %#v", result)
	}
	if len(result.CursorWarnings) != 1 || len(result.FutureRows) != 2 {
		t.Fatalf("diagnostics = %#v", result)
	}
}

func TestShortBudgetResumesAndRunnerCloseIsIdempotent(t *testing.T) {
	ctx := context.Background()
	path := filepath.Join(t.TempDir(), "skytrace.db")
	db, err := database.Open(ctx, path)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := db.SQL.Exec("INSERT INTO receivers (id) VALUES ('rx-1')"); err != nil {
		t.Fatal(err)
	}
	old := iso(time.Date(2025, 1, 1, 0, 0, 0, 0, time.UTC))
	transaction, err := db.SQL.Begin()
	if err != nil {
		t.Fatal(err)
	}
	statement, err := transaction.Prepare(`
		INSERT INTO track_points (
		  hex, receiver_id, observed_at, position_at, lat, lon
		) VALUES (?, 'rx-1', ?, ?, 37.5, 127.1)
	`)
	if err != nil {
		t.Fatal(err)
	}
	for index := 0; index < chunkSize+1; index++ {
		hex := fmt.Sprintf("%06x", index+1)
		if _, err := statement.Exec(hex, old, old); err != nil {
			t.Fatal(err)
		}
	}
	if err := statement.Close(); err != nil {
		t.Fatal(err)
	}
	if err := transaction.Commit(); err != nil {
		t.Fatal(err)
	}
	first, err := Run(ctx, db.SQL, time.Date(2026, 7, 30, 0, 0, 0, 0, time.UTC), Config{
		TrackRetentionDays: 90, BatchRetentionDays: 7, CurrentRetentionHours: 24,
		Budget: time.Nanosecond,
	})
	if err != nil {
		t.Fatal(err)
	}
	if first.Complete || first.TrackOld != chunkSize {
		t.Fatalf("first result = %#v", first)
	}
	second, err := Run(ctx, db.SQL, time.Date(2026, 7, 30, 0, 0, 0, 0, time.UTC), Config{
		TrackRetentionDays: 90, BatchRetentionDays: 7, CurrentRetentionHours: 24,
		Budget: time.Second,
	})
	if err != nil {
		t.Fatal(err)
	}
	if !second.Complete || second.TrackOld != 1 {
		t.Fatalf("second result = %#v", second)
	}
	if err := db.Close(); err != nil {
		t.Fatal(err)
	}

	runner, err := New(ctx, path, Config{FirstRun: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	if err := runner.Close(); err != nil {
		t.Fatal(err)
	}
	if err := runner.Close(); err != nil {
		t.Fatal(err)
	}
	if _, err := runner.RunNow(ctx); err == nil {
		t.Fatal("closed retention runner accepted a run")
	}
}
